clip_retrieval/clip_inference/worker.py

```diff
@@ -0,0 +1,123 @@
+from clip_retrieval.clip_inference.load_clip import load_clip
+from clip_retrieval.clip_inference.logger import LoggerReader, LoggerWriter
+from clip_retrieval.clip_inference.reader import FilesReader, WebdatasetReader
+
+
+def worker(
+    tasks,
+    input_dataset,
+    output_folder,
+    input_format="files",
+    cache_path=None,
+    batch_size=256,
+    num_prepro_workers=8,
+    enable_text=True,
+    enable_image=True,
+    enable_metadata=False,
+    wds_image_key="jpg",
+    wds_caption_key="txt",
+    clip_model="ViT-B/32",
+    mclip_model="sentence-transformers/clip-ViT-B-32-multilingual-v1",
+    use_mclip=False,
+    use_jit=True,
+    output_partition_count=None,
+    wandb_project="clip_retrieval",
+    enable_wandb=False,
+):
+    def reader_builder(sampler):
+        _, preprocess = load_clip(clip_model=clip_model, use_jit=use_jit)
+        if input_format == "files":
+            return FilesReader(
+                sampler,
+                preprocess,
+                input_dataset,
+                batch_size,
+                num_prepro_workers,
+                enable_text=enable_text,
+                enable_image=enable_image,
+                enable_metadata=enable_metadata,
+            )
+        elif input_format == "webdataset":
+            return WebdatasetReader(
+                sampler,
+                preprocess,
+                input_dataset,
+                batch_size,
+                num_prepro_workers,
+                enable_text=enable_text,
+                enable_image=enable_image,
+                enable_metadata=enable_metadata,
+                wds_image_key=wds_image_key,
+                wds_caption_key=wds_caption_key,
+                cache_path=cache_path,
+            )
+        else:
+            raise ValueError(f"Unknown input_format: {input_format}")
+
+    def mapper_builder():
+        return ClipMapper(
+            enable_image=enable_image,
+            enable_text=enable_text,
+            enable_metadata=enable_metadata,
+            use_mclip=use_mclip,
+            clip_model=clip_model,
+            use_jit=use_jit,
+            mclip_model=mclip_model,
+        )
+
+    def writer_builder(i):
+        return NumpyWriter(
+            partition_id=i,
+            output_folder=output_folder,
+            enable_text=enable_text,
+            enable_image=enable_image,
+            enable_metadata=enable_metadata,
+            output_partition_count=output_partition_count,
+        )
+
+    def logger_builder(i):
+        return LoggerWriter(
+            partition_id=i,
+            stats_folder=output_folder + "/stats",
+        )
+
+    runner = Runner(
+        reader_builder=reader_builder,
+        mapper_builder=mapper_builder,
+        writer_builder=writer_builder,
+        logger_builder=logger_builder,
+        output_partition_count=output_partition_count,
+    )
+
+    # TODO: figure out logging with SLURM
```

Comment:
This should be happening in the main, not in the worker

Reply:
how do you account for the logger needing to be passed to the `Reader`?

the slurm distributor will pass everything via CLI

Reply:
You create the logger reader in main
The logger writer in worker
Both of them take as input easy to serialize params